packages/core/src/runtime-feature.ts

```diff
@@ -0,0 +1,144 @@
+                runPromises.push(engine.runFeature(dep));
+            }
+            const envRunHandlers = this.runHandlers.get(envName);
+            if (envRunHandlers) {
+                for (const handler of envRunHandlers) {
+                    runPromises.push(handler());
+                }
+            }
+        }
+
+        await Promise.all(runPromises);
+    }
+}
+
+export function createFeatureRuntime<F extends FeatureDescriptor, E extends AnyEnvironment>(
+    feature: F,
+    runningEngine: RuntimeEngine<E>
+): RuntimeFeature<F, E> {
+    const { features, runOptions, referencedEnvs, entryEnvironment } = runningEngine;
+
+    const deps: any = {};
+    const depsApis: any = {};
+    const runningApi: any = {};
+    const inputApi: any = {};
+    const providedAPI: any = {};
+    const environmentContext: any = {};
+    const apiEntries = Object.entries(feature.api);
+
+    const setupHandlers = feature.runtimeInfo!.setups || [];
+    const contextHandlers = feature.runtimeInfo!.contexts || [];
+
+    for (const [key, api] of Object.entries(feature.api)) {
+        const entityFn = api[IDENTIFY_API];
+        if (entityFn) {
+            entityFn.call(api, feature.id, key);
+        }
+    }
+
+    const featureRuntime = new RuntimeFeature(feature, runningApi, deps, entryEnvironment);
+    features.set(feature.id, featureRuntime);
+
+    for (const dep of feature.dependencies) {
+        const instance = runningEngine.initFeature(dep);
+        deps[dep.id] = instance;
+        depsApis[dep.id] = instance.api;
+    }
+
+    for (const [key, entity] of apiEntries) {
+        const provided = entity[CREATE_RUNTIME](runningEngine, feature.id, key);
+        if (provided !== undefined) {
+            inputApi[key] = provided;
+        }
+    }
+    const settingUpFeature = {
+        ...inputApi,
+        id: feature.id,
+        run: featureRuntime.addRunHandler,
+        onDispose: featureRuntime.addOnDisposeHandler,
+        [RUN_OPTIONS]: runOptions,
+        [ENGINE]: runningEngine,
+        runningEnvironmentName: entryEnvironment.env,
+    };
+
+    for (const [key, contextHandler] of contextHandlers) {
+        environmentContext[key] = {
+            dispose: () => void 0,
+            ...contextHandler(depsApis),
+        };
+    }
+
+    for (const envName of referencedEnvs) {
+        const environmentSetupHandlers = setupHandlers.get(envName);
+        if (!environmentSetupHandlers) {
+            continue;
+        }
+        for (const setupHandler of environmentSetupHandlers) {
+            const featureOutput = setupHandler(settingUpFeature, depsApis, environmentContext);
+            if (!featureOutput) {
+                continue;
+            }
+            for (const key of Object.keys(featureOutput)) {
+                settingUpFeature[key] = featureOutput[key];
+            }
+            Object.assign(providedAPI, featureOutput);
+        }
+    }
+
+    for (const [key, entity] of apiEntries) {
```

Comment:
please extract to a named function, help the next dev to understand what this code is doing and why